docs/proposals/20200615-external-remediation.md

```diff
@@ -0,0 +1,299 @@
+
+No further action (deletion or applying annotations) will be taken by the MachineHealthCheck controller until the Node becomes healthy, when it will locate and delete the instantiated MachineRemediation CR.
+
+The ERC may wish to register a finalizer on it’s CR to ensure it has an opportunity to perform any additional cleanup in the case that the unhealthy state was transient and the Node returned to a healthy state prior to the completion of the full custom ERC flow.
+
+
+### User Stories
+
+#### Story 1
+As an admin of a hardware based cluster, I would like unhealthy nodes to be power-cycled, so that I can recover from transient errors faster and begin application recovery sooner.
+#### Story 2
+As an admin of a hardware based cluster, I would like unhealthy nodes to be power-cycled, so that I can detect hardware issues faster.
+#### Story 3
+As an admin of a hardware based cluster, I would like the system to keep attempting to power-cycle unhealthy nodes, so that they are automatically added back to the cluster when I fix the underlying problem.
+#### Story 4
+As an admin of a large bare metal cluster, I want to auto remediate nodes but do so grouped by failure domain (i.e. by rack). All nodes deemed unhealthy in a given rack should be cordoned first before each are drained.
+
+### Implementation Details/Notes/Constraints
+
+#### Flow
+The default MHC remediation strategy continues to be an annotation-based flow where the owning controller deletes the Machine (and by inference the Node) objects after performing any relevant cleanup.
+
+In order to enable a different remediation strategy for a set of Machines, the admin must provide a reference to a provider-specific template object as part of the MachineHealthCheck CR.
+
+When a Machine enters an unhealthy state, the MHC will:
+* Look up the referenced template
+* Instantiate the template (for simplicity, we will refer to this as a External Machine Remediation CR, or EMR)
+* Force the name and namespace to match the unhealthy Machine
+* Save the new object in etcd
+
+We use the same name and namespace for the External Machine Remediation CR to ensure uniqueness and lessen the possibility for multiple parallel remediations of the same Machine. 
+
+The lifespan of the EMRs is that of the remediation process and they are not intended to be a record of past events.  The EMR will also contain an ownerRef to the Machine, to ensure that it does not outlive the Machine it references.
+
+The only signaling between the MHC and the external controller watching EMR CRs is the creation and deletion of the EMR itself.  Any actions or changes that admin’s should be informed about should be emitted as events for consoles and UIs to consume if necessary.  They are informational only and do not result in or expect any behaviour from the MHC, Node, or Machine as a result.
```

Comment:
We can also use conditions for this signaling, this is the new standard we've been moving to.

```suggestion
The only signaling between the MHC and the external controller watching EMR CRs is the creation and deletion of the EMR itself.  Any actions or changes that admins should be informed about should be emitted as events for consoles and UIs to consume if necessary.  They are informational only and do not result in or expect any behaviour from the MHC, Node, or Machine as a result.
```